### PR TITLE
Allow configurable uuid for lvs creation 

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

This PR adds the ability to provide an optional `uuid` at the time of lvs create. 
If not provided one will be generated for it.

### CODE REVIEW

File: include/spdk/lvol.h [hunk 1]
```diff
@@ -73,6 +73,7 @@ struct spdk_lvs_opts {
 	uint32_t		cluster_sz;
 	enum lvs_clear_method	clear_method;
 	char			name[SPDK_LVS_NAME_MAX];
+	const char *uuid;
```
Comment (tiagolobocastro) [inline include/spdk/lvol.h:76]: I think formatting should be:
```C
char		name[SPDK_LVS_NAME_MAX];
const char      *uuid;
```
?
Is the linter running?

File: module/bdev/lvol/vbdev_lvol.c [hunk 2]
```diff
@@ -1379,3 +1379,74 @@ vbdev_lvol_create_with_uuid(struct spdk_lvol_store *lvs, const char *name, uint6
 
 	return rc;
 }
+
+int
+vbdev_lvs_create_with_uuid(const char *base_bdev_name, const char *name, const char *uuid,
```
Comment (tiagolobocastro) [inline module/bdev/lvol/vbdev_lvol.c:1384]: what about having the original create function call this one, with the uuid set to `NULL` to avoid code dup?

**Review by tiagolobocastro — APPROVED** [review]

File: include/spdk/lvol.h (hunk 1 above)
Comment (shubham14bajpai) [reply include/spdk/lvol.h]: I works but missed this I am not sure why

File: module/bdev/lvol/vbdev_lvol.c [hunk 3]
```diff
@@ -222,66 +222,9 @@ int
...
-	if (rc < 0) {
-		free(lvs_req);
-		bs_dev->destroy(bs_dev);
-		return rc;
-	}
-
-	return 0;
+	/* set uuid as NULL */
+	return vbdev_lvs_create_with_uuid(base_bdev_name, name, NULL, cluster_sz, clear_method, cb_fn,
+					  cb_arg);
```
Comment (jonathan-teh) [inline module/bdev/lvol/vbdev_lvol.c:227]: While this removes much of the code duplication there is still a large diff effectively moving the code elsewhere and will make upstream changes more difficult to merge in future. Rename the existing `vbdev_lvs_create()` to `vbdev_lvs_create_with_uuid()`, insert the code related to `uuid` there, and re-introduce the former function to call through to the latter.

File: shfmt.patch [hunk 4]
```diff
@@ -0,0 +1,87 @@
+--- scripts/setup.sh.orig
```
Comment (jonathan-teh) [inline shfmt.patch:1]: This file logically belongs to the .sh lint commit.

File: module/bdev/lvol/vbdev_lvol.c (hunk 3 above)
Comment (shubham14bajpai) [reply module/bdev/lvol/vbdev_lvol.c:227]: `vbdev_lvs_create()` cannot be called from `vbdev_lvs_create_with_uuid()` as the struct `spdk_lvs_opts` will be re initialized inside `vbdev_lvs_create()` and the uuid will be lost. I did not want to change the function signature of the existing function so IMO having a duplicate makes more sense here.

File: shfmt.patch (hunk 4 above)
Comment (shubham14bajpai) [reply shfmt.patch]: Oops

File: module/bdev/lvol/vbdev_lvol.c (hunk 3 above)
Comment (jonathan-teh) [reply module/bdev/lvol/vbdev_lvol.c:227]: No, call `vbdev_lvs_create_with_uuid()` from `vbdev_lvs_create()`, exactly as it is currently implemented but without moving the code around.

Comment (shubham14bajpai) [reply module/bdev/lvol/vbdev_lvol.c:227]: That would required adding `uuid` to `vbdev_lvs_create()` right?

Comment (shubham14bajpai) [reply module/bdev/lvol/vbdev_lvol.c:227]: Or changing the argument of `vbdev_lvs_create()` to accept `spdk_lvs_opts` instead of all different fields.

Comment (jonathan-teh) [reply module/bdev/lvol/vbdev_lvol.c:227]: Basically this:
```
diff --git a/module/bdev/lvol/vbdev_lvol.c b/module/bdev/lvol/vbdev_lvol.c
index 6f2c0c976..b72d929eb 100644
--- a/module/bdev/lvol/vbdev_lvol.c
+++ b/module/bdev/lvol/vbdev_lvol.c
@@ -219,6 +219,15 @@ end:
 int
 vbdev_lvs_create(const char *base_bdev_name, const char *name, uint32_t cluster_sz,
                 enum lvs_clear_method clear_method, spdk_lvs_op_with_handle_complete cb_fn, void *cb_arg)
+{
+       return vbdev_lvs_create_with_uuid(base_bdev_name, name, NULL, cluster_sz, clear_method, cb_fn,
+                                         cb_arg);
+}
+
+int
+vbdev_lvs_create_with_uuid(const char *base_bdev_name, const char *name,  const char *uuid,
+                uint32_t cluster_sz,
+                enum lvs_clear_method clear_method, spdk_lvs_op_with_handle_complete cb_fn, void *cb_arg)
 {
        struct spdk_bs_dev *bs_dev;
        struct spdk_lvs_with_handle_req *lvs_req;
@@ -240,6 +249,10 @@ vbdev_lvs_create(const char *base_bdev_name, const char *name, uint32_t cluster_
                opts.clear_method = clear_method;
        }
 
+       if (uuid != NULL) {
+               opts.uuid = uuid;
+       }
+
        if (name == NULL) {
                SPDK_ERRLOG("missing name param\n");
                return -EINVAL;
```

File: shfmt.patch (hunk 4 above)
Comment (jonathan-teh) [reply shfmt.patch]: This file is still present in the 1st commit and then later deleted in the 3rd commit. Please remove it from the 1st commit.

Comment (shubham14bajpai) [reply shfmt.patch]: Done

**Review by jonathan-teh — APPROVED** [review]

**Pull request merged**